libsignal-service/src/sender.rs

```diff
@@ -0,0 +1,151 @@
+}
+
+#[derive(serde::Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct SendMessageResponse {
+    pub needs_sync: bool,
+}
+
+/// Equivalent of Java's `SignalServiceMessageSender`.
+pub struct MessageSender<Service, Store> {
+    service: Service,
+    session_store: Store,
+    cipher: ServiceCipher,
+    device_id: i32,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum MessageSenderError {
+    #[error("Service error: {0}")]
+    ServiceError(#[from] ServiceError),
+    #[error("Protocol error: {0}")]
+    ProtocolError(#[from] libsignal_protocol::Error),
+}
+
+impl<Service, Store> MessageSender<Service, Store>
+where
+    Service: PushService,
+    Store: SessionStore,
+{
+    pub fn new(
+        service: Service,
+        session_store: Store,
+        cipher: ServiceCipher,
+        device_id: i32,
+    ) -> Self {
+        MessageSender {
+            service,
+            session_store,
+            cipher,
+            device_id,
+        }
+    }
+
+    /// Send a message (`content`) to an address (`recipient`).
+    // XXX: `online` supposedly has to do with Typing indicators.
+    // Cfr. libsignal-service-java 7eb925190d78360a1aaae13402b6eb747997aeca
+    pub async fn send_message(
+        &mut self,
+        recipient: crate::ServiceAddress,
+        content: impl Into<crate::content::ContentBody>,
+        timestamp: u64,
+        online: bool,
+    ) -> Result<(), MessageSenderError> {
+        let content = {
+            use prost::Message;
+            let content_proto = content.into().into_proto();
+            let mut content = Vec::with_capacity(content_proto.encoded_len());
+            content_proto
+                .encode(&mut content)
+                .expect("infallible message encoding");
+            content
+        };
+
+        let messages =
+            self.create_encrypted_messages(&recipient, None, &content)?;
+        let messages = OutgoingPushMessages {
+            destination: recipient.get_identifier(),
+            timestamp,
+            messages,
+            online,
+        };
+        self.service.send_messages(messages).await?;
+        Ok(())
+    }
+
+    // Equivalent with `getEncryptedMessages`
+    fn create_encrypted_messages(
+        &mut self,
+        recipient: &crate::ServiceAddress,
+        unidentified_access: Option<bool>,
+        content: &[u8],
+    ) -> Result<Vec<OutgoingPushMessage>, MessageSenderError> {
+        let mut messages = vec![];
+
+        let myself = *recipient == self.cipher.local;
+        if !myself
+            || (self.device_id != DEFAULT_DEVICE_ID
+                || unidentified_access.is_some())
+        {
+            let message = self.create_encrypted_message(
+                recipient,
+                unidentified_access,
+                DEFAULT_DEVICE_ID,
+                content,
+            )?;
+            messages.push(message);
+        }
+
+        // TODO: get sub-device-sessions here to send to myself!
```

Comment:
I think I'd still like this in this PR; the way I debug is with Signal Desktop, which is sending to myself :-)

Reply:
Just added it, it works now! I've tested with "Note to self".